### DISCUSSION TITLE
[BUG] type casting while using Dict and Union, and it varies with the order inside Union

### DISCUSSION THREAD

**Discussion by ali-lqc** [fastapi/fastapi#9818] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# Example 1 --------------- 
from fastapi import FastAPI
from pydantic import BaseModel
import uvicorn
from typing import Dict, Union, Any

app = FastAPI()


class MyData(BaseModel):
    # try commenting/ uncommenting each of the lines below to see the bug
    dataset1: Dict
    # dataset1: Dict[str, Any]
    # dataset1: Dict[str, Union[str, int, float]]
    # dataset1: Dict[str, Union[int, float, str]]
    # dataset1: Dict[str, Union[float, int, str]]


@app.post("/process_data")
def process_data(data: MyData):
    return data


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)


# Example 2 --------------- 
from fastapi import FastAPI
from pydantic import BaseModel
import uvicorn
from typing import Dict, Union, Any

app = FastAPI()


class MyData(BaseModel):
    dataset1: Dict[str, Union[str, int, float]]
    dataset2: Dict[str, Union[int, float, str]]
    dataset3: Dict[str, Union[float, int, str]]


@app.post("/process_data")
def process_data(data: MyData):
    return data


if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

### Description

# Example 1
This is referring to example 1 in the code.
when sending the curl request below:
`curl -X POST -H "Content-Type: application/json" -d "{"dataset1":{"A":1,"B":"Bee","D":4.5}}"  http://localhost:8000/process_data `

as you can notice, A is an int, B is a string, and C is a float.
the response types vary depending on the typing:

1. Dict returns -> `{"dataset1":{"A":1,"B":"Bee","D":4.5}}`
2. Dict[str, Any] returns -> `{"dataset1":{"A":1,"B":"Bee","D":4.5}}`
3. Dict[str, Union[str, int, float]] returns ->` {"dataset1":{"A":"1","B":"Bee","D":"4.5"}}`
4. Dict[str, Union[int, float, str]] returns -> `{"dataset1":{"A":1,"B":"Bee","D":4}}`
5. Dict[str, Union[float, int, str]] returns -> `{"dataset1":{"A":1.0,"B":"Bee","D":4.5}}`

As you can see, no 1 and no 2 are correct, however, no 3 return all strings, no 4 casted float from 4.5 to int 4, and no 5 casted the int 1 to float 1.0

# Example 2
This is referring to example 2 in the code.
when sending the curl request below, with all three datasets having identical values:
`curl -X POST -H "Content-Type: application/json" -d "{"dataset1":{"A":1,"B":"Bee","D":4.5}, "dataset2":{"A":1,"B":"Bee","D":4.5}, "dataset3":{"A":1,"B":"Bee","D":4.5}}"  http://localhost:8000/process_data `

the response is below:
`{"dataset1":{"A":"1","B":"Bee","D":"4.5"},"dataset2":{"A":"1","B":"Bee","D":"4.5"},"dataset3":{"A":"1","B":"Bee","D":"4.5"}}`

There appears to be some cache effect happening in addition to the bug shown in example 1.
Because dataset1 had str type appear first, dataset2 and dataset3 were all casted to string.
However, if the order was changed as below:
`dataset2: Dict[str, Union[int, float, str]]`
`dataset1: Dict[str, Union[str, int, float]]`
`dataset3: Dict[str, Union[float, int, str]]`
the response casts the floats to integers in all 3 cases:
`{"dataset2":{"A":1,"B":"Bee","D":4},"dataset1":{"A":1,"B":"Bee","D":4},"dataset3":{"A":1,"B":"Bee","D":4}}`

### Operating System

Linux, Windows

### FastAPI Version

0.64.0 & 0.99.1

### Python Version

Python 3.7.11 & Python 3.10

**Comment by ybressler** [comment]
The error you are facing has to do with discriminated unions – more specifically, pydantic is automatically converting some of the values into designated formats.
* [V1 Conversion table](https://docs.pydantic.dev/1.10/usage/models/#data-conversion)
* [V2 Conversion table](https://docs.pydantic.dev/latest/usage/conversion_table/)

When you perform a union, Pydantic is going to consider any transformation which doesn't raise an error as valid. For example: `Union[str, int, float]` will convert `1` as `'1'` since the `integer 1` can be correctly transformed to a `string`.

Automatic conversion behavior changes in V2. But the automatic conversion still occurs.

**Further reading:**
You may want to look into [discriminated unions](https://docs.pydantic.dev/1.10/usage/types/#discriminated-unions-aka-tagged-unions).

**Reply by ali-lqc** [reply]
Thank you for your answer, but the problem also happens without using pydantic at all, as shown in the example below:
Note: that this conversion doesn't happen when using Union with a variable directly, such as:
`dataset: Union[str, int, float]`
it's only happening when used in combination with Dict, such as:
`dataset: Dict[str, Union[str, int, float]]`


Example that doesn't use pydantic:
```
from fastapi import FastAPI
import uvicorn
from typing import Dict, Union, Any

app = FastAPI()

@app.post("/process_data")
def process_data(data: Dict[str, Union[str, int, float]]):
    return data

if __name__ == "__main__":
    uvicorn.run(app, host="0.0.0.0", port=8000)
```

`curl -X POST -H "Content-Type: application/json" -d "{"A":1,"B":"Bee","D":4.5}"  http://localhost:8000/process_data`

the response is:
`{"A":"1","B":"Bee","D":"4.5"}`

Everything has been converted to string

**Reply by ybressler** [reply]
Let me correct myself. Pydantic is wrapping [Python's `union`](https://docs.python.org/3/library/typing.html#typing.Union). What happens if you call the `process_data()` function directly? I imagine it should return the same input as provided.

However, when you submit data to the endpoint, `FastApi` is validating your data via Pydantic, so you will get the coerced data.

**Comment by YuriiMotov** [comment, marked as answer]
Just to put all information together in one answer:
* input data is validated by Pydantic even if you don't use models
* when Pydantic validates `Union`, it takes the first type that passes the validation. So, if the first type of your `Union` is `str` it always wins
* you should put types in specific order from most strict to less strict
* you may want to use strict types such as `StrictFloat` and `StrictInt` to prevent coercing data
